TensorLib/TensorData.lean

```diff
@@ -0,0 +1,226 @@
+/-
+Copyright (c) 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: John Tristan, Paul Govereau, Sean McLaughlin
+-/
+import Mathlib.Tactic
+import TensorLib.NonEmptyList
+import TensorLib.TensorElement
+
+namespace TensorLib
+
+/-!
+This file defines TensorData, the raw data underlying a Tensor.
+We are representing it today as nested linked lists. We predict
+this will be easier to reason about, but will be slow to evaluate.
+We intend to have a more compact and faster-to-access representation
+in addition to this one in the future for execution.
+-/
+
+-- I tried making this implicit with {} but I needed to add @ everwhere
+-- and give it the argument anyway. Maybe there is a way to infer at least
+-- within a section
+variable (a: Type)
+
+inductive TensorData where
+| td_root (xs: List a)
+| td_node (xs: NonEmptyList TensorData)
+
+instance : Inhabited (TensorData a) where
+  default := .td_root []
+
+def td_node! (xs: List (TensorData a)): TensorData a :=
```

Comment:
We need to make a decision. Do we allow ourselves to use "!" functions? I think it may be a mistake.

Reply:
This was just to make the examples that follow without `do` notation. I'll make it private. I'm not actually sure what to do with the examples. Seems like they should be tests or something. Will look what other people do.